base64/decode_arm64.s

```diff
@@ -0,0 +1,203 @@
+	VTBL    V25.B8, [V2.B8], V25.B8;                             \
+	VTBL    V18.B8, [V3.B8], V26.B8;                             \
+	VTBL    V19.B8, [V3.B8], V27.B8;                             \
+	VTBL    V20.B8, [V3.B8], V28.B8;                             \
+	VTBL    V21.B8, [V3.B8], V29.B8;                             \
+	VAND    V22.B8, V26.B8, V26.B8;                              \
+	VAND    V23.B8, V27.B8, V27.B8;                              \
+	VAND    V24.B8, V28.B8, V28.B8;                              \
+	VAND    V25.B8, V29.B8, V29.B8;                              \
+	WORD    $0x0e209b5a /* VCMEQ   $0, V26.B8, V26.B8 */;        \
+	WORD    $0x0e209b7b /* VCMEQ   $0, V27.B8, V27.B8 */;        \
+	WORD    $0x0e209b9c /* VCMEQ   $0, V28.B8, V28.B8 */;        \
+	WORD    $0x0e209bbd /* VCMEQ   $0, V29.B8, V29.B8 */;        \
+	VORR    V26.B8, V27.B8, V26.B8;                              \
+	VORR    V28.B8, V29.B8, V28.B8;                              \
+	VORR    V26.B8, V28.B8, V26.B8;                              \
+	VMOV    V26.D[0], R5;                                        \
+	VMOV    V26.D[1], R6;                                        \
+	ORR     R6, R5;                                              \
+	CBNZ    R5, goto_err;                                        \
+	/* Shift hi nibles */                                        \
+	VTBL    V18.B8, [V4.B8], V18.B8;                             \
+	VTBL    V19.B8, [V4.B8], V19.B8;                             \
+	VTBL    V20.B8, [V4.B8], V20.B8;                             \
+	VTBL    V21.B8, [V4.B8], V21.B8;                             \
+	VBIT    V14.B8, V8.B8, V18.B8;                               \
+	VBIT    V15.B8, V8.B8, V19.B8;                               \
+	VBIT    V16.B8, V8.B8, V20.B8;                               \
+	VBIT    V17.B8, V8.B8, V21.B8;                               \
+	/* Combine results */                                        \
+	VADD    V18.B8, V10.B8, V10.B8;                              \
+	VADD    V19.B8, V11.B8, V11.B8;                              \
+	VADD    V20.B8, V12.B8, V12.B8;                              \
+	VADD    V21.B8, V13.B8, V13.B8;                              \
+	VUSHR   $4, V11.B8, V14.B8;                                  \
+	VUSHR   $2, V12.B8, V15.B8;                                  \
+	VSHL    $2, V10.B8, V10.B8;                                  \
+	VSHL    $4, V11.B8, V11.B8;                                  \
+	VSHL    $6, V12.B8, V12.B8;                                  \
+	VORR    V10.B8, V14.B8, V16.B8;                              \
+	VORR    V11.B8, V15.B8, V17.B8;                              \
+	VORR    V12.B8, V13.B8, V18.B8
+
+#define ADVANCE_LOOP(goto_loop)                                \
+	VST3.P  [V16.B8, V17.B8, V18.B8], 24(R3);                    \
+	ADD     $32, R4;                                             \
+	CMP     R4, R2;                                              \
+	BGT     goto_loop
```

Comment:
Could this be `BGE` or do we need more than 32 bytes of input to proceed with another iteration?

Reply:
Ah it does move by 32 bytes, but I'm advancing the pointer here (`R4`) and comparing it to a pre-calculated end pointer (`R2'). The end is calculated to be one past the last byte:

https://github.com/segmentio/asm/blob/484fd1ffa11393e879d7a4214c30a53815e10588/base64/decode_arm64.s#L9